src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -158,58 +164,49 @@ LoweringToDatalogPass::BaseFactToDLIR(const Principal& speaker,
 
 std::pair<datalog::Predicate, std::vector<datalog::Rule>>
 LoweringToDatalogPass::FactToDLIR(const Principal& speaker, const Fact& fact) {
-  return std::visit(
-      raksha::utils::overloaded{
-          [this, speaker](const BaseFact& base_fact) {
-            return BaseFactToDLIR(speaker, base_fact);
-          },
-          [this, speaker](const std::unique_ptr<CanSay>& can_say) {
-            auto [inner_fact_dlir, gen_rules] =
-                FactToDLIR(speaker, *can_say->fact());
-
-            Principal fresh_principal(FreshVar());
-
-            // The following code generates the extra rule that does
-            // delegation. This rule is:
-            // ```
-            // speaker says inner_fact_dlir:-
-            //      fresh_principal says inner_fact_dlir,
-            //      speaker says fresh_principal canSay inner_fact_dlir
-            // ```
-
-            // This is p says inner_fact_dlir
-            auto lhs = PushPrincipal("says_", speaker, inner_fact_dlir);
-            auto fresh_prin_says_inner =
-                PushPrincipal("says_", fresh_principal, inner_fact_dlir);
-            auto speaker_says_fresh_cansay_inner = PushPrincipal(
-                "says_", speaker,
-                PushPrincipal("canSay_", fresh_principal, inner_fact_dlir));
-            auto rhs = {fresh_prin_says_inner, speaker_says_fresh_cansay_inner};
-            gen_rules.push_back(datalog::Rule(lhs, rhs));
-
-            // Note that prin_cansay_pred does not begin with "speaker says"
-            // because only the top-level fact should. This could
-            // be a recursive call, so it might not be processing
-            // the top-level fact. The top-level fact gets wrapped
-            // in a "says" during the call to translate the assertion
-            // in which this appears.
-            auto prin_cansay_pred =
-                PushPrincipal("canSay_", can_say->principal(), inner_fact_dlir);
-            return std::make_pair(prin_cansay_pred, gen_rules);
-          }},
-      fact.GetValue());
+  auto [inner_fact_dlir, gen_rules] = BaseFactToDLIR(speaker, fact.basefact());
+  for (const CanSay& can_say : fact.cansay()) {
+    Principal fresh_principal(FreshVar());
+    // The following code generates the extra rule that does
+    // delegation. This rule is:
+    // ```
+    // speaker says inner_fact_dlir:-
+    //      fresh_principal says inner_fact_dlir,
+    //      speaker says fresh_principal canSay inner_fact_dlir
+    // ```
+
+    // This is p says inner_fact_dlir
+    auto lhs = PushPrincipal("says_", speaker, inner_fact_dlir);
+    auto fresh_prin_says_inner =
+        PushPrincipal("says_", fresh_principal, inner_fact_dlir);
+    auto speaker_says_fresh_cansay_inner = PushPrincipal(
+        "says_", speaker,
+        PushPrincipal("canSay_", fresh_principal, inner_fact_dlir));
+    auto rhs = {fresh_prin_says_inner, speaker_says_fresh_cansay_inner};
```

Comment:
What is the type of `rhs`? I genuinely don't know. I think `auto` should be replaced with an explicit type annotation here.

Reply:
Added types. rhs is of type datalog::Predicate